client/src/worldgen.rs

```diff
@@ -40,83 +46,551 @@ impl NodeState {
-        match self {
-            Hole(_) | RootSky | Sky | DeepSky => Some(Material::Void),
-            DeepLand => Some(Material::Stone),
-            _ => None,
+pub fn voxels(graph: &mut DualGraph, node: NodeId, cube: Vertex) -> VoxelData {
+    let max_elevations = chunk_incident_max_elevations(graph, node, cube);
+
+    let mut voxels = VoxelData::Uninitialized;
+
+    for ([x, y, z], path) in cube.dual_vertices() {
+        let state = graph
+            .get(path.fold(node, |node, side| graph.neighbor(node, side).unwrap()))
+            .as_ref()
+            .unwrap();
+        let subchunk_offset = na::Vector3::new(x as usize, y as usize, z as usize);
+        state.write_subchunk(&mut voxels, subchunk_offset, cube, max_elevations);
+    }
+
+    voxels
+}
+
+fn chunk_incident_max_elevations(graph: &DualGraph, node: NodeId, cube: Vertex) -> [f64; 8] {
+    let mut e = cube
+        .dual_vertices()
+        .map(|(_, path)| path.fold(node, |node, side| graph.neighbor(node, side).unwrap()))
+        .map(|n| graph.get(n).as_ref().unwrap().max_elevation as f64);
+
+    // this is a bit cursed, but I don't want to collect into a vec because perf,
+    // and I can't just return an iterator because then something still references graph.
+    let e1 = e.next().unwrap();
+    let e5 = e.next().unwrap();
+    let e3 = e.next().unwrap();
+    let e7 = e.next().unwrap();
+    let e2 = e.next().unwrap();
+    let e6 = e.next().unwrap();
+    let e4 = e.next().unwrap();
+    let e8 = e.next().unwrap();
+
+    [e1, e2, e3, e4, e5, e6, e7, e8]
+}
+
+/// Keeps track of the canonical surface wrt. the NodeState this is stored in
+pub struct Surface {
+    normal: na::Vector4<f64>,
+}
+impl Surface {
+    /// A Vector pointing up from the surface at the root node.
+    fn at_root() -> Self {
+        let n = Side::A.reflection().map(|x| x as f64) * origin() - origin();
+        Self {
+            normal: lorentz_normalize(&n),
+        }
+    }
+
+    /// Returns a new Surface relative to a node on a certain side of this one
+    fn reflect(&self, side: Side) -> Self {
+        Self {
+            normal: lorentz_normalize(&(side.reflection() * self.normal)),
         }
     }
 
-    pub fn write_chunk(self, voxels: &mut VoxelData, subchunk_offset: na::Vector3<usize>) {
-        if let Some(material) = self.solid_material() {
-            match voxels {
-                // who knows, maybe this voxel is surrounded by the same type of node.
-                VoxelData::Uninitialized => {
-                    *voxels = VoxelData::Solid(material);
-                    return;
+    /// The max_elevation of a single voxel wrt. this Surface
+    fn voxel_elevation(&self, voxel: na::Vector3<f64>, cube: Vertex) -> f64 {
+        let pos = lorentz_normalize(&(cube.cube_to_node() * voxel.push(1.0)));
+        mip(&pos, &self.normal).asinh()
+    }
+}
+
+fn trilerp<N: na::RealField>(a: &[N], t: na::Vector3<N>) -> N {
```

Comment:
It's strange for this to have a dynamic length. Maybe take advantage of the `TryFrom` impls for array references from slices?